File: frontend/src/api/organize.ts
import { request, requestJson } from './client'

export type OrganizeMode = 'keep' | 'date' | 'camera'
export type OrganizeStatus = 'pending' | 'running' | 'completed' | 'failed'

export interface OrganizeRequest {
  folders: string[]
  destination: string
  mode: OrganizeMode
  rename: boolean
  skip_duplicates: boolean
}

export interface RenameExample {
  old: string
  new: string
}

export interface OrganizePreview {
  total: number
  planned: number
  duplicates_in_set: number
  duplicates_skipped: number
  already_organized: number
  undated: number
  est_bytes: number
  example_paths: string[]
  rename_example: RenameExample | null
  // Destination is outside the indexed folders; organizing will add it
  // automatically so the moved photos stay in the Library.
  destination_new_root: boolean
  // Destination sits inside a folder being organized — photos would move into
  // a subfolder of themselves.
  destination_inside_source: boolean
}

/** One place the library currently lives. A date tree reports the destination
 *  it was organized into, not each of its year/month folders. */
export interface LibraryLocation {
  path: string
  photos: number
}

export interface LibraryLayout {
  locations: LibraryLocation[]
  total: number
}

export interface OrganizeRun {
  id: number
  status: OrganizeStatus
  params: OrganizeRequest
  batch_id: string
  total: number
  planned: number
  moved: number
  skipped_duplicates: number
  already_organized: number
  undated: number
  failed_count: number
  est_bytes: number
  message: string | null
  started_at: string | null
  finished_at: string | null
  created_at: string
}

export const TERMINAL_ORGANIZE_STATUSES: OrganizeStatus[] = ['completed', 'failed']

export function previewOrganize(req: OrganizeRequest): Promise<OrganizePreview> {
  return requestJson<OrganizePreview>('/api/organize/preview', 'POST', req)
}

export function getLibraryLayout(): Promise<LibraryLayout> {
  return request<LibraryLayout>('/api/organize/layout')
}

export function startOrganize(req: OrganizeRequest): Promise<OrganizeRun> {
  return requestJson<OrganizeRun>('/api/organize', 'POST', req)
}

export function getOrganizeRun(id: number): Promise<OrganizeRun> {
  return request<OrganizeRun>(`/api/organize/${id}`)
}

export function listOrganizeRuns(limit = 1): Promise<OrganizeRun[]> {
  return request<OrganizeRun[]>(`/api/organize?limit=${limit}`)
}

File: backend/app/schemas/organize.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

OrganizeMode = Literal["keep", "date", "camera"]
OrganizeStatus = Literal["pending", "running", "completed", "failed"]


class OrganizeRequest(BaseModel):
    folders: list[str] = Field(min_length=1)
    destination: str = Field(min_length=1)
    mode: OrganizeMode
    rename: bool = False
    skip_duplicates: bool = True


class RenameExample(BaseModel):
    old: str
    new: str


class LibraryLocation(BaseModel):
    path: str
    photos: int


class LibraryLayout(BaseModel):
    """Where organized photos actually live, so a second location is visible
    before a run creates one rather than after."""

    locations: list[LibraryLocation]
    total: int


class OrganizePreviewRead(BaseModel):
    total: int
    planned: int
    duplicates_in_set: int
    duplicates_skipped: int
    already_organized: int
    undated: int
    est_bytes: int
    # First few planned destinations, one per distinct target directory.
    example_paths: list[str]
    rename_example: RenameExample | None
    # Destination is outside every scan root today; starting the run will
    # register it automatically so the moved photos stay in the Library.
    destination_new_root: bool
    # Destination sits inside a folder being organized — photos would move into
    # a subfolder of themselves. Surfaced as a warning, never a refusal.
    destination_inside_source: bool


class OrganizeRunRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    status: OrganizeStatus
    params: OrganizeRequest
    batch_id: str
    total: int
    planned: int
    moved: int
    skipped_duplicates: int
    already_organized: int
    undated: int
    failed_count: int
    est_bytes: int
    message: str | None
    started_at: datetime | None
    finished_at: datetime | None
    created_at: datetime
